src/cluster_legacy.c

```diff
@@ -2990,7 +2991,46 @@ int clusterProcessPacket(clusterLink *link) {
 
                 if (clusterNodeIsMaster(sender)) {
                     /* Master turned into a slave! Reconfigure the node. */
-                    clusterDelNodeSlots(sender);
+                    if (master && !memcmp(master->shard_id, sender->shard_id, CLUSTER_NAMELEN)) {
+                        /* `sender` was a primary and was in the same shard as `master`, its new primary */
+                        if (sender->configEpoch > senderConfigEpoch) {
+                            serverLog(LL_NOTICE,
+                                    "Ignore stale message from %.40s (%s) in shard %.40s;"
+                                    " gossip config epoch: %llu, current config epoch: %llu", 
+                                    sender->name,
+                                    sender->human_nodename,
+                                    sender->shard_id,
+                                    (unsigned long long)senderConfigEpoch,
+                                    (unsigned long long)sender->configEpoch);
+                        } else {
+                            /* A failover occurred in the shard where `sender` belongs to and `sender` is no longer
+                             * a primary. Update slot assignment to `master`, which is the new primary in the shard */
+                            int slots = clusterMoveNodeSlots(sender, master);
```

Comment:
The `master` is still a replica in this observer node's view, but it will start owning slots.

1. Do we want to update the role of `master` to primary in our view?
2. Should we bump the `master`'s epoch to `senderConfigEpoch`?

Reply:
Thanks @srgsanky. 

I think you have a good point. `master` is indeed a replica in this observer's view. And if we make it primary, we do need to update its epoch to `senderConfigEpoch` too. Let me update the fix next.

Reply:
lgtm